KAFKA-14402: Update AddPartitionsToTxn protocol to batch and handle verifyOnly requests

jolshan opened the pull request:
Part 1 of [KIP-890]( https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)

I've updated the API spec and related classes.

Clients should only be able to send up to version 3 requests and that is enforced by using a client builder.

Requests > 4 only require cluster permissions as they are initiated from other brokers.

I've added tests for the batched requests and for the verifyOnly mode.

Also -- minor change to the KafkaApis method to properly match the request name.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jolshan committed:
basic batching

jolshan committed:
cleaner batching

jolshan committed:
bug fixes

jolshan committed:
add verifyOnly and auth changes

jolshan committed:
more tests

jolshan committed:
fixes

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+        public final boolean isClientRequest;
 
-        public Builder(final AddPartitionsToTxnRequestData data) {
+        // Only used for versions < 4
```
does this mean this constructor is only used for less than 4? how are we enforcing this?

jolshan replied:
Maybe Jason's suggestion below will help -- using factory methods.
I'm not sure there is any way to "enforce" but to guide folks to use the right methods

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+            AddPartitionsToTxnTopicCollection topics = compileTopics(partitions);
+
+            this.data = new AddPartitionsToTxnRequestData()
+                    .setTransactionalId(transactionalId)
+                    .setProducerId(producerId)
+                    .setProducerEpoch(producerEpoch)
+                    .setTopics(topics);
```
the indentation looks off here. i think intellij does this to me too

jolshan replied:
Are you just saying I should do 4 instead of 8? I didn't know if there was a formalized decision there

jeffkbkim replied:
i don't know either 😅 i only bring this up because i had feedback on this before. it seems like other places use both double indentation and single so maybe we should have a formalized decision

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+            this.isClientRequest = false;
 
+            this.data = new AddPartitionsToTxnRequestData()
+                    .setTransactions(transactions)
```
same on the indentation

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -101,15 +130,61 @@ public String toString() {
+    }
+    
+    public Map<String, List<TopicPartition>> partitionsByTransaction() {
+        if (cachedPartitionsByTransaction != null && cachedPartitionsByTransaction.size() == data.transactions().size()) {
```
can there be any case where the two maps have the same size but store different items?

jolshan replied:
I'm not sure how that would happen

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -101,15 +130,61 @@ public String toString() {
         return cachedPartitions;
     }
+    
+    private List<TopicPartition> partitionsForTransaction(String transaction) {
```
are we using this return type anywhere?

jolshan replied:
I'm not sure I understand the question.

jeffkbkim replied:
sorry, i meant the return value `List<TopicPartition>`, it doesn't seem to be used

jolshan replied:
?? are you saying the method isn't used? It's used on line 168

jeffkbkim replied:
the method can be of type void, no?

jolshan replied:
not as it is written. We return on line 150. I originally thought this could be useful if we just wanted the list of the partitions. but since it is private I can change.

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -49,28 +52,37 @@ public class AddPartitionsToTxnResponse extends AbstractResponse {
+    private static AddPartitionsToTxnTopicResultCollection topicCollectionForErrors(Map<TopicPartition, Errors> errors) {
         Map<String, AddPartitionsToTxnPartitionResultCollection> resultMap = new HashMap<>();
-
+        
```
nit: maybe just a newline?

jeffkbkim commented on clients/src/test/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponseTest.java:
```diff
@@ -84,16 +88,59 @@ public void testParse() {
-                                                  .setThrottleTimeMs(throttleTimeMs);
-        AddPartitionsToTxnResponse response = new AddPartitionsToTxnResponse(data);
-
         for (short version : ApiKeys.ADD_PARTITIONS_TO_TXN.allVersions()) {
```
can we use the parameterized test for the api versions?

jolshan replied:
I thought maybe we didn't want to redo the top steps every time but sure i can change it.

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2383,68 +2384,101 @@ class KafkaApis(val requestChannel: RequestChannel,
-        else
-          authorizedPartitions.add(topicPartition)
+    val lock = new Object
+    val addPartitionsToTxnRequest = if (request.context.apiVersion() < 4) request.body[AddPartitionsToTxnRequest].normalizeRequest() else request.body[AddPartitionsToTxnRequest]
```
nit: i think we can separate the if else here

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2383,68 +2384,101 @@ class KafkaApis(val requestChannel: RequestChannel,
+    def maybeSendResponse(): Unit = {
+      lock synchronized {
+        if (responses.size() == txns.size()) {
+          requestHelper.sendResponseMaybeThrottle(request, createResponse)
```
do we need to hold the lock when sending the response?

jolshan replied:
no

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2383,68 +2384,101 @@ class KafkaApis(val requestChannel: RequestChannel,
+    }
+
+    txns.forEach( transaction => {
+      val transactionalId = transaction.transactionalId()
```
nit: parentheses

jeffkbkim commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -352,7 +353,12 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
               Left(Errors.NONE)
             } else {
-              Right(coordinatorEpoch, txnMetadata.prepareAddPartitions(partitions.toSet, time.milliseconds()))
+              // If verifyOnly, we should have returned in the step above. If we didn't the partitions are not present in the transaction.
```
does this comment meant that if the request was to verifyOnly, then we should have returned with Left(Errors.NONE) because there should have been a matching transaction with all of the partitions?

jolshan replied:
yes

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2383,68 +2384,101 @@ class KafkaApis(val requestChannel: RequestChannel,
-          nonExistingTopicErrors += topicPartition -> Errors.UNKNOWN_TOPIC_OR_PARTITION
-        else
-          authorizedPartitions.add(topicPartition)
+    val lock = new Object
```
any reason we don't use `val responses` as the synchronization object?

jolshan replied:
i can do that

jeffkbkim commented on core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala:
```diff
@@ -302,17 +302,42 @@ class TransactionCoordinatorTest {
     )
   }
-
+  
```
nit: indentation

jeffkbkim commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -66,24 +97,22 @@ public Builder(final String transactionalId,
                 topics.add(new AddPartitionsToTxnTopic()
-                               .setName(partitionEntry.getKey())
-                               .setPartitions(partitionEntry.getValue()));
+                        .setName(partitionEntry.getKey())
+                        .setPartitions(partitionEntry.getValue()));
```
i'm seeing a bunch of places with double indentation - not sure if it's intentional or it's done by your editor

jolshan replied:
Are you saying it shouldn't be indented here? I removed 8 spaces.

jeffkbkim reviewed (commented):
Thanks for the PR! left some comments

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+            AddPartitionsToTxnTopicCollection topics = compileTopics(partitions);
+
+            this.data = new AddPartitionsToTxnRequestData()
+                    .setTransactionalId(transactionalId)
```
Does it make sense to set verifyOnly to false explicitly here?

jolshan replied:
We can do that in the "fromClient" factory method 👍

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+                    .setVerifyOnly(verifyOnly);
+        }
+
+        private AddPartitionsToTxnTopicCollection compileTopics(final List<TopicPartition> partitions) {
```
nit: how about `buildTxnTopicCollection`?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -66,24 +97,22 @@ public Builder(final String transactionalId,
         @Override
         public AddPartitionsToTxnRequest build(short version) {
-            return new AddPartitionsToTxnRequest(data, version);
+            short clampedVersion = (isClientRequest && version > 3) ? 3 : version;
```
It's a little strange to ignore the version. I think another way to do this is to set the `latestAllowedVersion` to 3 in the client builder. That will ensure that the client does not try to use a higher version even if the broker supports it. Similarly, we can set a min version of 4 for the server.

jolshan replied:
Hmm I didn't see such an option in the builder but maybe I'm missing something.

hachikuji replied:
It comes from `AbstractRequest.Builder`. You can pass the min and max versions in the constructor.

jolshan replied:
I didn't see if we directly built the request for the AddPartitionsToTxn requst or passed the abstract request to a queue (where we lose the type 😅 ) I will check again though

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -101,15 +130,61 @@ public String toString() {
         return cachedPartitions;
     }
+    
+    private List<TopicPartition> partitionsForTransaction(String transaction) {
```
nit: could we use `transactionalId` as the argument name?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -118,11 +193,41 @@ public AddPartitionsToTxnRequestData data() {
+            }
+            return new AddPartitionsToTxnResponse(throttleTimeMs, errors);
+        } else {
+            AddPartitionsToTxnResponseData response = new AddPartitionsToTxnResponseData();
```
The logic in here makes me wonder if we should add a top-level error code in the response.

jolshan replied:
There a few errors we can return as top level errors in code and that's for auth to the transaction (client side only), ensuring ibp, and cluster auth (for server side only)  if that helps. The rest are topic or partition level.

hachikuji replied:
Many new APIs do include a top-level error code, so there is definitely precedent. I kind of like giving the broker an easy way to indicate a failure without requiring it to do a bunch of wasteful work. It's not a common case though, so I'm not sure how much it matters.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -35,21 +44,43 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+        public final boolean isClientRequest;
 
-        public Builder(final AddPartitionsToTxnRequestData data) {
+        // Only used for versions < 4
```
Perhaps we could create factory methods: `Builder.forClient` and `Builder.forServer` (or something like that).

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -49,28 +52,37 @@ public class AddPartitionsToTxnResponse extends AbstractResponse {
         this.data = data;
     }
 
+    // Only used for versions < 4
```
I do wonder if we need both constructors. The old version is a subset of the new version where the number of producers happens to be 1.

jolshan replied:
I left it this way for minimal code changes but we could get rid of this one.

hachikuji replied:
Yeah, the problem is that it makes the usage more confusing. I'd rather have a larger patch I guess and try to abstract some of the version details away.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -99,6 +112,7 @@ public void maybeSetThrottleTimeMs(int throttleTimeMs) {
         data.setThrottleTimeMs(throttleTimeMs);
     }
 
+    // Only used for versions < 4
```
Do we need to continue exposing some of these version-specific methods? We sort of intend the request/response objects to abstract a lot of the version-specific details.

jolshan replied:
I'm not sure how to get around some of these since the return type is different.

hachikuji replied:
As mentioned elsewhere, the old API is a special case of the new API where there just happens to be a single transactionalId. I think if we take that perspective, then we can make the methods general and hide the version details.

jolshan replied:
I guess if we want to deal with a map or larger data structure that has only one object we can. I think the tricky thing is knowing when we expect just one response how to use it correctly. That may require further refactoring but it is probably doable.

hachikuji commented on clients/src/main/resources/common/message/AddPartitionsToTxnResponse.json:
```diff
@@ -22,22 +22,35 @@
+      { "name": "TopicResults", "type": "[]AddPartitionsToTxnTopicResult", "versions": "4+",
+        "about": "The results for each topic." }
+    ]},
+    { "name": "Results", "type": "[]AddPartitionsToTxnTopicResult", "versions": "0-3",
```
I wonder if it makes sense to use a different naming convention for the fields which are being removed. For example, instead of `Results` we could have `ResultsV3AndBelow`. It is not pretty, but it does make the usage in the code clearer.

jolshan replied:
That's fair. We've done something similar in the past.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -352,7 +353,12 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
-              Right(coordinatorEpoch, txnMetadata.prepareAddPartitions(partitions.toSet, time.milliseconds()))
+              // If verifyOnly, we should have returned in the step above. If we didn't the partitions are not present in the transaction.
+              if (verifyOnly) {
+                Left(Errors.INVALID_TXN_STATE)
```
Suppose that some of the partitions are added to the transaction and some are not. We return `INVALID_TXN_STATE` which means the broker cannot distinguish which partitions have been added correctly. Does it matter? I think there are two main consequences:
1. We have to reject the full `Produce` request, which may contain writes to multiple partitions. This means the producer also cannot tell which partition was not added correctly. Maybe this is fine since the producer must abort anyway?
2. The `AddPartitionsToTxn` cannot batch across multiple `Produce` requests from the same `transactionalId`. Or, if it does batch across requests, then we would have to return INVALID_TXN_STATE in all of the responses. Maybe this is also fine for the same reason?

I do think we'll want to have at least in the logs somewhere which partitions were not present in the transaction.

jolshan replied:
You bring up good points and I can give more detailed response on partition level.

jolshan replied:
I think this was mostly an oversight on my part.

guozhangwang replied:
+1 here, since we do have the finest error-code level at the per-partition level we can leverage it to be more detailed.

As for whether we should also add a global error code: in general I'd like that idea, and I think there are a few error codes: auth, unsupported_version, unsupported_format (we also have a `UNSUPPORTED_FOR_MESSAGE_FORMAT` for that), that should always be global to save the sender time to dig into lower-leveled ones. And then for different request type there are some global errors based on just the global fields --- for this case, the only global field left now is `verifyOnly` so that we probably do not have additional error codes. I vaguely remember there's a KIP proposing such a principle but I cannot find the KIP now.

If we agree this is a generally good idea then let's start with this one first :P

jolshan replied:
Ack -- we may make verifyOnly not a global field, but the point still stands.

guozhangwang commented on clients/src/test/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequestTest.java:
```diff
@@ -17,43 +17,138 @@
-        List<TopicPartition> partitions = new ArrayList<>();
-        partitions.add(new TopicPartition("topic", 0));
-        partitions.add(new TopicPartition("topic", 1));
+        
```
Do we already have a test coverage with the old `Builder` constructor to check that `verifyOnly` would default to `false`?

jolshan replied:
We don't but it's easy to add

guozhangwang replied:
thx!

guozhangwang commented on clients/src/main/resources/common/message/AddPartitionsToTxnResponse.json:
```diff
@@ -22,22 +22,35 @@
+      { "name": "TopicResults", "type": "[]AddPartitionsToTxnTopicResult", "versions": "4+",
+        "about": "The results for each topic." }
+    ]},
+    { "name": "Results", "type": "[]AddPartitionsToTxnTopicResult", "versions": "0-3",
```
nit: maybe rename this field from `Results` to `TransactionResultsByTopic`?

guozhangwang commented on clients/src/main/resources/common/message/AddPartitionsToTxnResponse.json:
```diff
@@ -22,22 +22,35 @@
-        { "name": "ErrorCode", "type": "int16", "versions": "0+",
-          "about": "The response error code."}
-      ]}
+      { "name": "Results", "type": "[]AddPartitionsToTxnPartitionResult", "versions": "0+",
```
ditto here, maybe `TopicResultsByPartition`?

dajac commented on clients/src/main/resources/common/message/AddPartitionsToTxnRequest.json:
```diff
@@ -23,17 +23,35 @@
   "flexibleVersions": "3+",
   "fields": [
-    { "name": "TransactionalId", "type": "string", "versions": "0+", "entityType": "transactionalId",
+    { "name": "VerifyOnly", "type": "bool", "versions": "4+", "default": false,
```
I was wondering if having `VerifyOnly` as top level field is the right thing to do. It basically means that we cannot batch transactions with `VerifyOnly=true` and transactions with `VerifyOnly=false` in the same request. This may impact out batching mechanism. Thoughts?

jolshan replied:
Is the idea that we will have some old producers (verifyOnly) and some new ones?

jolshan replied:
One benefit is that verifyOnly requests won't be slowed down by the non-verify only ones

dajac replied:
Right. At the broker level, I think that it is possible to have old and new producers, right? If we use one connection from the broker to the transaction coordinator, it means that one has to wait on the other. I don't know how you plan to implement this so it is just a thought.

jolshan replied:
It's a fair point. I'll think on it a bit.

hachikuji replied:
Yeah, interesting point. I could see moving `VerifyOnly` to the level of `transactionalId`. An interesting corollary if we do batch both modes together is that verify-only requests may end up blocking on replication even though we are only checking the state in memory. This would kind of penalize old clients, but maybe that's acceptable. Unless we used separate connections for each mode, perhaps it is unavoidable.

jolshan replied:
Yup that was my point here:
> One benefit is that verifyOnly requests won't be slowed down by the non-verify only ones

But yeah, I guess we slow things down as well by using separate requests. So maybe unavoidable

hachikuji replied:
Yep, you got there first 😄. I would probably vote to make the change. Tracking separate connections does not sound attractive. We could still do that even with the modified protocol if there is a noticeable regression for old clients.

jolshan replied:
I'm wondering do we want to keep the top level verifyOnly as an optimization? Or is it just more trouble than it's worth and maybe won't be used too often.

jolshan replied:
Actually -- the way this is implemented, we check on a per-transaction basis anyway. Since this is 1:1 corresponded to producer ID (and therefore client version) the only time we could see changes between verify only and not would be when the client upgrades. I may need to think about handling that as a special case.

hachikuji commented on clients/src/main/resources/common/message/AddPartitionsToTxnRequest.json:
```diff
@@ -23,17 +23,35 @@
-  "validVersions": "0-3",
+  //
+  // Version 4 adds VerifyOnly field to check if partitions are already in transaction and adds support to batch multiple transactions.
+  "validVersions": "0-4",
```
Would it make sense to set `latestVersionUnstable` so that we reserve some flexibility to change the API after we merge this?

jolshan replied:
Yeah makes sense

jolshan committed:
update request json

jolshan committed:
new builders

jolshan committed:
Fix up requests/responses

jolshan committed:
Per Partition error codes for verify only

jolshan commented:
I've added the changes to the API spec
-- verify only is now a transaction level config
-- top level error is added to the response

I've added builders to the request and tried to simplify some of the methods.
I've also addressed the verifyOnly case where some partitions are in the txn and others are not. 

I will update the KIP to reflect some of these changes (especially with respect to the API spec)

I still need to address the unstable API change, but that will require a pull from master.

jolshan committed:
Merge branch 'trunk' of github.com:apache/kafka into kafka-14402

jolshan commented:
Took a quick look at the unstable api change. Looks like some integration tests built specifically for v4 fail with `org.apache.kafka.common.errors.InvalidRequestException: Received request api key ADD_PARTITIONS_TO_TXN with version 4 which is not enabled` 

I will need to look into this.

dajac commented:
> Took a quick look at the unstable api change. Looks like some integration tests built specifically for v4 fail with `org.apache.kafka.common.errors.InvalidRequestException: Received request api key ADD_PARTITIONS_TO_TXN with version 4 which is not enabled`
> 
> I will need to look into this.

@jolshan I suppose that you have to enable unstable apis in your new integration tests.

jolshan committed:
unstable apis support

jolshan committed:
Fix tests

jolshan committed:
just change the config for the test

guozhangwang reviewed (approved):
Made a new pass according to the newly updated KIP, do not have further comments.

guozhangwang commented:
@dajac @hachikuji if you do not have further comments, we can proceed and merge it then?

dajac commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1328,7 +1328,7 @@ Priority priority() {
         public void handleResponse(AbstractResponse response) {
             AddPartitionsToTxnResponse addPartitionsToTxnResponse = (AddPartitionsToTxnResponse) response;
-            Map<TopicPartition, Errors> errors = addPartitionsToTxnResponse.errors();
+            Map<TopicPartition, Errors> errors = addPartitionsToTxnResponse.errors().get(AddPartitionsToTxnResponse.V3_AND_BELOW_TXN_ID);
```
nit: I suppose that `errors` should never be `null` here. I wonder if we should still check it. What do you think?

jolshan replied:
What should we do if the check fails? Just have a better error message thrown?

dajac replied:
Yeah, the check is probably not necessary. By the way, I find the idea of having `V3_AND_BELOW_TXN_ID` for old version a bit confusing. I was wondering if using `addPartitionsToTxnResponse.data().resultsByTopicV3AndBelow()` would be a better alternative here. We only iterate over the Map so the Map is not strictly required here. Have you considered something like this?

jolshan replied:
I was told not to have v3 and below specific methods from Jason because the v3 case should generalize to a single version of the v4 case and that should make it easy to use methods for both.

However, if we really think this is an issue. I guess we can change the approach again. I'm just not sure the experience of errors only applying to v4+. Any ideas there besides changing the method name to express it should only be used in v4+?

dajac replied:
Understood. Let's keep it as it is then.

I agree that v3 case should generalized to a single item of the v4 case. It is just unfortunate that we don't have the transaction id in v3 response so we have to use an empty string for it. I suppose that it is the way it is.

jolshan replied:
Yeah. It really is unfortunate. 😞

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -34,22 +43,40 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+                    .setV3AndBelowTransactionalId(transactionalId)
+                    .setV3AndBelowProducerId(producerId)
+                    .setV3AndBelowProducerEpoch(producerEpoch)
+                    .setV3AndBelowTopics(topics));
```
nit: The indentation of the arguments looks inconsistent.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -34,22 +43,40 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
     private final AddPartitionsToTxnRequestData data;
 
-    private List<TopicPartition> cachedPartitions = null;
+    private final short version;
```
The version is already in the base class. Do we really need it here?

dajac commented on clients/src/main/resources/common/message/AddPartitionsToTxnRequest.json:
```diff
@@ -23,17 +23,39 @@
   // Version 3 enables flexible versions.
-  "validVersions": "0-3",
+  //
+  // Version 4 adds VerifyOnly field to check if partitions are already in transaction and adds support to batch multiple transactions.
```
I think that we should explain that v4 is only for other brokers and clients are suppose to use version <= v3.

dajac commented on clients/src/main/resources/common/message/AddPartitionsToTxnRequest.json:
```diff
@@ -23,17 +23,39 @@
+      { "name": "Topics", "type": "[]AddPartitionsToTxnTopic", "versions": "4+",
+        "about": "The partitions to add to the transaction." }
+    ]},
+    { "name": "V3AndBelowTransactionalId", "type": "string", "versions": "0-3", "entityType": "transactionalId",
```
Is it common to prefix old fields by their version? It is the first time I see it.

jolshan replied:
I had previous comments from @hachikuji and @guozhangwang that I should do something like this to avoid confusion when setting the fields. I have seen something similiar in other places but we didn't explicitly say the version.

dajac replied:
Ack. That seems fine.

dajac commented on clients/src/main/resources/common/message/AddPartitionsToTxnResponse.json:
```diff
@@ -22,22 +22,37 @@
       "about": "Duration in milliseconds for which the request was throttled due to a quota violation, or zero if the request did not violate any quota." },
-    { "name": "Results", "type": "[]AddPartitionsToTxnTopicResult", "versions": "0+",
-      "about": "The results for each topic.", "fields": [
+    { "name": "ErrorCode", "type": "int16", "versions": "4+",
```
Should we make it ignorable in case we would set it by mistake?

jolshan replied:
Sounds fine to me.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -118,11 +123,78 @@ public AddPartitionsToTxnRequestData data() {
+        if (version < 4) {
+            response.setResultsByTopicV3AndBelow(errorResponseForTopics(data.v3AndBelowTopics(), error));
+        } else {
+            AddPartitionsToTxnResultCollection results = new AddPartitionsToTxnResultCollection();
+            for (AddPartitionsToTxnTransaction transaction : data().transactions()) {
+                results.add(errorResponseForTransaction(transaction.transactionalId(), error));
+            }
+            response.setResultsByTransaction(results);
+            response.setErrorCode(error.code());
```
When there is a global error, do we expect to set both the top level error and to return an error for each transaction? If we alway set both, what is the purpose of the top level error?

jolshan replied:
As discussed by @hachikuji and @guozhangwang the idea was if a top level error was set we could skip the rest of the handling. I just set all the fields to the same error for consistency. What is your suggestion? To remove these

dajac replied:
Yeah, I like the top level error code for that reason. However, we must be clear on how we want to use it. I suppose that we can only use it for cluster authorization failure and unexpected errors failing the entire request/response.

jolshan replied:
To clarify -- I'm not sure if being clear on how we want to use it (ie the errors it is used for) is the same question as if we want to include the error on the txns as well. 

As for the latter, I think we are more error prone if we set the error to none. I guess I'm just not sure what we are trying to accomplish by removing them from the lower fields. To be clear, I think I've also seen this pattern on other responses with top level errors.

dajac replied:
I was looking at the AlterPartition API as a reference. There, when we have a top level error, we don't set the partitions in the response at all. We basically save space. See [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/AlterPartitionRequest.java#L46).

jolshan replied:
Ok.
I will remove the errors from the response
When using the new API it is mandatory I check the top level error before any further handling. 
I may also need to adjust tests/test this functionality.

dajac replied:
Don't forget to take the version into account here. We can only do this for v4.

jolshan replied:
Yup. This will only be done in v4 responses. So for now, we don't do any checks since the server side usage is upcoming in the next PR.

As a side note, seems that AlterPartition includes the top level error in errorCounts so I did that here as well.

dajac commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -1303,11 +1305,13 @@ public void testCommitWithTopicAuthorizationFailureInAddPartitionsInFlight() thr
         client.respond(body -> {
             AddPartitionsToTxnRequest request = (AddPartitionsToTxnRequest) body;
-            assertEquals(new HashSet<>(request.partitions()), new HashSet<>(errors.keySet()));
+            assertEquals(new HashSet<>(AddPartitionsToTxnRequest.getPartitions(request.data().v3AndBelowTopics())), new HashSet<>(errors.keySet()));
```
`AddPartitionsToTxnRequest.getPartitions(request.data().v3AndBelowTopics()))` looks a bit weird from an encapsulation perspective. Why not just keeping `partitions` as before if you need it?

jolshan replied:
I was told to avoid having version specific methods. In order to use by both the v3 and below, and v4 objects this was the best way to do it. It really only looks bad in the tests.

dajac replied:
I tend to agree with this. Are those helpers only used in tests? If so, it may be better to put them in `AddPartitionsToTxnRequestTest`. What do you think?

jolshan replied:
I can take a look. I also used internally to build the request and I think the v4+ version in KafkaApis itself. I can try to make a helper for these tests though.

dajac replied:
If those are used internally the request as well, we can't really move them.

jolshan replied:
The one in the request is used in a place that iterates through the topic objects so it is actually ok. I made a helper in the TransactionManagerTest since it uses the phrase 3 times. I left the final remaining usage of this in the last test file since I think its ok to have once and doesn't make sense to use a helper.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+      if (version < 4) {
+        // There will only be one response in data. Add it to the response data object.
+        val data = new AddPartitionsToTxnResponseData()
+        responses.forEach(result => {
```
nit: `responses.forEach(result => {` -> `responses.forEach { result => `

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+    }
+
+    txns.forEach( transaction => {
```
nit: `txns.forEach( transaction => {` -> `txns.forEach { transaction =>`.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+        val nonExistingTopicErrors = mutable.Map[TopicPartition, Errors]()
+        val authorizedPartitions = mutable.Set[TopicPartition]()
+
+        val authorizedTopics = if (version < 4) authHelper.filterByAuthorized(request.context, WRITE, TOPIC,
+          partitionsToAdd.filterNot(tp => Topic.isInternal(tp.topic)))(_.topic) else partitionsToAdd.map(_.topic).toSet
```
nit: I find this line quite hard to read. Should we put it on multiple lines?

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+            transaction.producerEpoch,
+            authorizedPartitions,
+            transaction.verifyOnly,
+            sendResponseCallback,
+            sendVerifyResponseCallback,
```
I am not sure to understand why we need two callbacks. I find it weird in the first place. My understanding is that in both cases, we end up with an error per partition in the response so it seems to me that we could unify them, no?

jolshan replied:
It is not the same. In one case we have a single error for all topic partitions, but in the other, we may have different errors per partition. (If we verify and have some that exist and others that don't)

jolshan replied:
I had to keep the single error because this callback is used in the append path and I didn't want to change it.

dajac replied:
> It is not the same. In one case we have a single error for all topic partitions, but in the other, we may have different errors per partition. (If we verify and have some that exist and others that don't)

Yeah, I understand this. My point was that on the non-verify case, we also end up having an error per partition in the response (the same one for all of them).

> I had to keep the single error because this callback is used in the append path and I didn't want to change it.

Hum... I think that we could handle this in `handleAddPartitionsToTransaction` without changing the append path. Basically, when an error is return by the append, we could build a map with the error for all partitions.

I think that having two callbacks is a bad design choice here but that's just my humble opinion.

jolshan replied:
Is the request to make a mapping of every topic partition to the same error? That doesn't seem super efficient, but I guess it makes things simpler in the parameters. We would also need to define a new callback in `handleAddPartitionsToTransaction` that calls the original callback (so we have two still)

I suppose this is doable though.

jolshan replied:
I've changed to one callback, but I decided to go with a different approach where we either provide one error or the map. I think it would be annoying to convert the errors for all partitions and then potentially have to redo it to convert the error for v2 and below. Let me know if there is an issue with this.

dajac replied:
I am not sure whether I like it or not. Those two values in the callback may be error prone in the future. Reading `handleAddPartitionsToTransaction` again, the add and verify steps only have the code validation in common. How feasible would it be to actually add another method for the verification part instead of trying to do everything in `handleAddPartitionsToTransaction`? I suppose that we could extract the code part into a common method used by both.

jolshan replied:
I suppose I can do that but it looked like a lot of code duplication. There wasn't a great way to extract out the commonalities but I can try again.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+          }
+          
+          def sendVerifyResponseCallback(errors: Map[TopicPartition, Errors]): Unit = {
+            responses.synchronized {
+              responses.add(AddPartitionsToTxnResponse.resultForTransaction(transactionalId, errors.asJava))
+            }
+            maybeSendResponse()
```
This pattern is used in a few places. Would having `addResponseAndMaybeSend` helper method make sense?

jolshan replied:
Maybe. But the parameters may be tricky. 😅 The maybeSendResponse was supposed to be the helper. But I guess it's not helpful enough.

jolshan replied:
I guess I can just supply what to add to the response. I will redo this.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
-          new AddPartitionsToTxnResponse(requestThrottleMs, partitionErrors.asJava))
+    val txns = addPartitionsToTxnRequest.data.transactions
+    def maybeSendResponse(): Unit = {
+      var canSend = false
+      responses.synchronized {
+        if (responses.size() == txns.size()) {
+          canSend = true
+        }
+      }
```
nit: A more scala-ish version of this:

```
val canSend = responses.synchronized {
  responses.size == txns.size
}
```

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2493,7 +2542,9 @@ class KafkaApis(val requestChannel: RequestChannel,
         Set(offsetTopicPartition),
+        false,
         sendResponseCallback,
+        null,
```
This tend to confirm that having two callbacks is weird.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
-      .build()
+    val request =
+      if (version < 4) {
+        AddPartitionsToTxnRequest.Builder.forClient(
+          transactionalId,
+          producerId,
+          producerEpoch,
+          List(createdTopicPartition, nonExistentTopic).asJava)
+          .build()
```
nit: `build()` looks weird here. Should we put the closing parenthesis of `forClient` on a new line: `).build()`?

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+      if (version < 4) 
+        response.errors.get(AddPartitionsToTxnResponse.V3_AND_BELOW_TXN_ID) 
+      else 
+        response.errorsForTransaction(response.getTransactionTopicResults(transactionalId))
```
nit: `response.errorsForTransaction(response.getTransactionTopicResults(transactionalId))` looks really weird from an encapsulation perspective. Can't we use `response.errors.get(transactionalId)`?

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+    
+    val txn2Topics = new AddPartitionsToTxnTopicCollection()
+    txn2Topics.add(new AddPartitionsToTxnTopic()
+      .setName(tp0.topic())
+      .setPartitions(Collections.singletonList(tp0.partition())))
```
nit: `()` are not necessary. There are few similar cases in this file.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+
+    val errors = response.errors()
+
+    assertTrue(errors.containsKey(transactionalId1))
+    assertTrue(errors.get(transactionalId1).containsKey(tp0))
+    assertEquals(Errors.NONE, errors.get(transactionalId1).get(tp0))
+
+    assertTrue(errors.containsKey(transactionalId2))
+    assertTrue(errors.get(transactionalId1).containsKey(tp0))
+    assertEquals(Errors.INVALID_PRODUCER_ID_MAPPING, errors.get(transactionalId2).get(tp0))
```
nit: How about using `assertEquals(Map(... define the expect map ..., errors)`? This is simpler and has the benefit of ensuring that we only have what we expect in `errors`.

Note that assertion as L148 is not correct. `transactionalId2` should be used.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+
+    val verifyErrors = verifyResponse.errors()
+
+    assertTrue(verifyErrors.containsKey(transactionalId))
+    assertTrue(verifyErrors.get(transactionalId).containsKey(tp0))
+    assertEquals(Errors.INVALID_TXN_STATE, verifyErrors.get(transactionalId).get(tp0))
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+    val producerEpoch1 = initPidResponse.data().producerEpoch()
+
+    val txn1Topics = new AddPartitionsToTxnTopicCollection()
+    partitions.foreach(tp => 
```
nit: `partitions.foreach {`

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+object AddPartitionsToTxnRequestServerTest {
+   def parameters: JStream[Arguments] = {
+    val arguments = mutable.ListBuffer[Arguments]()
+    ApiKeys.ADD_PARTITIONS_TO_TXN.allVersions().forEach( version =>
```
nit: `forEach { `.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+   def parameters: JStream[Arguments] = {
+    val arguments = mutable.ListBuffer[Arguments]()
+    ApiKeys.ADD_PARTITIONS_TO_TXN.allVersions().forEach( version =>
+      Array("kraft", "zk").foreach( quorum =>
```
nit: ditto.

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -1962,7 +1962,9 @@ class KafkaApisTest {
         ArgumentMatchers.eq(producerId),
         ArgumentMatchers.eq(epoch),
         ArgumentMatchers.eq(Set(new TopicPartition(Topic.GROUP_METADATA_TOPIC_NAME, partition))),
+        ArgumentMatchers.eq(false),
```
It would be great if we could also add new unit tests to cover the batch mode in `KafkaApisTest`.

jolshan replied:
I added tests in RequestServerTest but I guess those are more integration tests. I can add some in KafkaApisTest too.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -48,29 +51,50 @@ public class AddPartitionsToTxnResponse extends AbstractResponse {
+
+        for (AddPartitionsToTxnResult result : this.data.resultsByTransaction()) {
+            String transactionalId = result.transactionalId();
+            errorsMap.put(transactionalId, errorsForTransaction(data().resultsByTransaction().find(transactionalId).topicResults()));
```
Can't you reuse `result` instead of calling `data().resultsByTransaction().find(transactionalId)`?

dajac reviewed (commented):
@jolshan Thanks for the PR. I made a first pass and left some comments.

jolshan committed:
Nits, changed to one callback, added kafka apis test for batched mode

jolshan committed:
Redo callbacks by splitting into two methods, clean up error handling for top level errors, minor changes

jolshan committed:
test fix

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -34,22 +43,38 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+            AddPartitionsToTxnTopicCollection topics = buildTxnTopicCollection(partitions);
+            
+            return new Builder(ApiKeys.ADD_PARTITIONS_TO_TXN.oldestVersion(),
+                (short) 3, 
```
nit: Should we put `(short) 3` on the previous line to be consistent with how you did it at L66?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -34,22 +43,38 @@ public class AddPartitionsToTxnRequest extends AbstractRequest {
+                    .setTransactions(transactions));
+        }
+        
+        public Builder(short minVersion, short maxVersion, AddPartitionsToTxnRequestData data) {
```
nit: Do we still use this constructor anywhere? It may be good to make it private or package private to ensure that `forClient` or `forBroker` is used.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java:
```diff
@@ -103,26 +113,82 @@ public AddPartitionsToTxnRequest(final AddPartitionsToTxnRequestData data, short
+        if (version() < 4) {
+            response.setResultsByTopicV3AndBelow(errorResponseForTopics(data.v3AndBelowTopics(), error));
+        } else {
+            AddPartitionsToTxnResultCollection results = new AddPartitionsToTxnResultCollection();
+            response.setResultsByTransaction(results);
```
nit: I think that you can remove those two lines. The collection should be initialized by default.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -48,29 +51,51 @@ public class AddPartitionsToTxnResponse extends AbstractResponse {
+    public Map<String, Map<TopicPartition, Errors>> errors() {
+        Map<String, Map<TopicPartition, Errors>> errorsMap = new HashMap<>();
+
+        if (this.data.resultsByTopicV3AndBelow().size() != 0) {
```
nit: I think that we usually prefer using `isEmpty()`.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -80,45 +105,44 @@ topicName, new AddPartitionsToTxnPartitionResultCollection()
+        List<Errors> allErrors = new ArrayList<>();
+
+        // If we are not using this field, we have request 4 or later
+        if (this.data.resultsByTopicV3AndBelow().size() == 0) {
```
nit: `isEmpty()`?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponse.java:
```diff
@@ -80,45 +105,44 @@ topicName, new AddPartitionsToTxnPartitionResultCollection()
+
+        // If we are not using this field, we have request 4 or later
+        if (this.data.resultsByTopicV3AndBelow().size() == 0) {
+            allErrors.add(Errors.forCode(data.errorCode()));
```
nit: Should we use `updateErrorCounts` from `AbstractResponse` instead of creating `allErrors`?

jolshan replied:
I create allErrors because I use addAll for the individual transactions. I can place this code after I create error counts, but it doesn't really seem to accomplish much.

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/AddPartitionsToTxnResponseTest.java:
```diff
@@ -58,42 +65,75 @@ public void setUp() {
+        
+        assertEquals(txn1Errors, errorsForTransaction(response.getTransactionTopicResults("txn1")));
+        assertEquals(txn2Errors, errorsForTransaction(response.getTransactionTopicResults("txn2")));
+    }
```
nit: Should we add a test for `errors()`?

jolshan replied:
I tested in other tests, but I can put one here.

jolshan replied:
(ErrorCounts also uses this method)

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -2598,12 +2604,37 @@ private OffsetsForLeaderEpochResponse createLeaderEpochResponse() {
+                    singletonList(new TopicPartition("topic", 73))).build(version);
+        } else {
+            AddPartitionsToTxnTransactionCollection transactions = new AddPartitionsToTxnTransactionCollection(
+                    singletonList(new AddPartitionsToTxnTransaction()
```
nit: Indentation of this line looks weird. Should it be on the previous line?

jolshan replied:
It's a bit too long for the previous line though. I made it 4 instead of 8 spaces.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -317,6 +322,34 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+      
+      result match {
+        case Left(err) =>
+          debug(s"Returning $err error code to client for $transactionalId's AddPartitions request")
```
nit: Should update this line to mention that we are validating only here?

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -317,6 +322,34 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+
+        case Right((_, txnMetadata)) =>
+          val txnMetadataPartitions = txnMetadata.topicPartitions
+          val addedPartitions = partitions.intersect(txnMetadataPartitions)
+          val nonAddedPartitions = partitions.diff(txnMetadataPartitions)
+          val errors = mutable.Map[TopicPartition, Errors]()
+          addedPartitions.foreach(errors.put(_, Errors.NONE))
+          nonAddedPartitions.foreach(errors.put(_, Errors.INVALID_TXN_STATE))
```
nit: I am not sure if it makes a real difference but did you consider doing something like this:

```
partitions.foreach { tp =>
  if (txnMetadata.topicPartitions.contains(tp))
    ...
  else
   ...
}
```

If works, it would avoid allocating the intermediate collections. I leave this up to you.

jolshan replied:
sure

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -330,44 +363,53 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
   }
+  
+  private def getTransactionMetadata(transactionalId: String,
+                             producerId: Long,
+                             producerEpoch: Short,
+                             partitions: collection.Set[TopicPartition]): ApiResult[(Int, TransactionMetadata)] = {
```
nit: Indentation is off.

dajac commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -330,44 +363,53 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
+                             producerId: Long,
+                             producerEpoch: Short,
+                             partitions: collection.Set[TopicPartition]): ApiResult[(Int, TransactionMetadata)] = {
+    // try to update the transaction metadata and append the updated metadata to txn log;
+    // if there is no such metadata treat it as invalid producerId mapping error.
```
nit: Is this comment relevant here?

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,66 +2386,111 @@ class KafkaApis(val requestChannel: RequestChannel,
+    def addResultAndMaybeSendResponse(result: AddPartitionsToTxnResult): Unit = {
+      val canSend = responses.synchronized {
+        responses.add(result)
+        responses.size() == txns.size()
```
nit: You can remove the `()`.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2384,66 +2386,111 @@ class KafkaApis(val requestChannel: RequestChannel,
-          authorizedPartitions,
-          sendResponseCallback,
-          requestLocal)
+          if (!transaction.verifyOnly) {
+            txnCoordinator.handleAddPartitionsToTransaction(transactionalId,
+              transaction.producerId,
+              transaction.producerEpoch,
+              authorizedPartitions,
+              sendResponseCallback,
+              requestLocal)
+          } else {
+            txnCoordinator.handleVerifyPartitionsInTransaction(transactionalId,
+              transaction.producerId,
+              transaction.producerEpoch,
+              authorizedPartitions,
+              addResultAndMaybeSendResponse)
+          }
+        }
```
👍🏻

dajac commented on core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala:
```diff
@@ -314,6 +316,32 @@ class TransactionCoordinatorTest {
+    when(transactionManager.getTransactionState(ArgumentMatchers.eq(transactionalId)))
+      .thenReturn(Right(Some(CoordinatorEpochAndTxnMetadata(coordinatorEpoch,
+        new TransactionMetadata(transactionalId, 0, 0, 0, RecordBatch.NO_PRODUCER_EPOCH, 0, Empty, partitions, 0, 0)))))
+
```
nit: Extra empty line.

dajac commented on core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala:
```diff
@@ -1185,4 +1213,8 @@ class TransactionCoordinatorTest {
   }
+
+  def verifyPartitionsInTxnCallback(result: AddPartitionsToTxnResult): Unit = {
+    errors = AddPartitionsToTxnResponse.errorsForTransaction(result.topicResults()).asScala.toMap
```
Relying on a global variable is risky here. It would be much better to define the callback within the test itself.

jolshan replied:
I'm just followed the pattern of the errorsCallback. 😅 I'm not going to redo that, but I can redo this if it's an issue.

dajac commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala:
```diff
@@ -55,22 +65,146 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
+          producerId,
+          producerEpoch,
+          List(createdTopicPartition, nonExistentTopic).asJava
+        ).build()
```
Should we set the version here?

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2030,13 +2032,93 @@ class KafkaApisTest {
   }
 
+  @Test
+  def testBatchedRequest(): Unit = {
```
nit: Could we include the related api in the name?

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2030,13 +2032,93 @@ class KafkaApisTest {
+
+    createKafkaApis().handleAddPartitionsToTxnRequest(request, requestLocal)
+
+    verify(requestChannel).sendResponse(
+      ArgumentMatchers.eq(request),
+      capturedResponse.capture(),
+      ArgumentMatchers.eq(None)
+    )
+    val response = capturedResponse.getValue
```
nit: You can use `verifyNoThrottling`.

dajac reviewed (commented):
@jolshan Thanks for the update. I left some nits. We should be good to go once they are addressed.

jolshan committed:
Small changes

dajac commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -231,7 +231,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
     ApiKeys.INIT_PRODUCER_ID -> ((resp: InitProducerIdResponse) => resp.error),
     ApiKeys.WRITE_TXN_MARKERS -> ((resp: WriteTxnMarkersResponse) => resp.errorsByProducerId.get(producerId).get(tp)),
-    ApiKeys.ADD_PARTITIONS_TO_TXN -> ((resp: AddPartitionsToTxnResponse) => resp.errors.get(tp)),
+    ApiKeys.ADD_PARTITIONS_TO_TXN -> ((resp: AddPartitionsToTxnResponse) => resp.errors.get(AddPartitionsToTxnResponse.V3_AND_BELOW_TXN_ID).get(tp)),
```
As a follow-up: We should cover the new version here as well.

jolshan replied:
The new version doesn't really use authorizer, so I wasn't sure if it was needed.

jolshan replied:
I guess it uses `authHelper.authorizeClusterOperation(request, CLUSTER_ACTION)`

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2030,13 +2032,87 @@ class KafkaApisTest {
   }
 
+  @Test
+  def testBatchedAddPartitionsToTxnRequest(): Unit = {
```
As a follow-up: It seems that the test coverage is pretty low for this API here. It would be great if we could extend it. e.g. authorization failures, validation failures, etc.

dajac reviewed (approved):
LGTM, thanks for the patch. I left two suggestions for follow-ups. We can do those in separate PRs. This one is large enough.

jolshan commented:
for follow ups: https://issues.apache.org/jira/browse/KAFKA-14790
